File: cluster-cli/src/main/scala/org/midonet/cluster/cli/commands/objects/Interface.scala
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.cluster.cli.commands.objects

import org.midonet.cluster.data.{ZoomObject, ZoomClass, ZoomField}
import org.midonet.cluster.models.Topology
import org.midonet.cluster.util.IPAddressUtil.{Converter => IPAddressConverter}
import org.midonet.cluster.util.MACUtil.{Converter => MACConverter}
import org.midonet.packets.{IPAddr, MAC}

@ZoomClass(clazz = classOf[Topology.Host.Interface])
@CliName(name = "interface")
class Interface extends ZoomObject with Obj {

    @ZoomField(name = "name")
    @CliName(name = "name")
    var name: String = _
    @ZoomField(name = "type")
    @CliName(name = "type", readonly = false)
    var interfaceType: InterfaceType = _
    @ZoomField(name = "mac", converter = classOf[MACConverter])
    @CliName(name = "mac", readonly = false)
    var mac: MAC = _
    @ZoomField(name = "addresses", converter = classOf[IPAddressConverter])
    @CliName(name = "addresses")
    var addresses: Set[IPAddr] = Set.empty
    @ZoomField(name = "up")
    @CliName(name = "up", readonly = false)
    var isUp: Boolean = _
    @ZoomField(name = "has_link")
    @CliName(name = "has-link", readonly = false)
    var hasLink: Boolean = _
    @ZoomField(name = "mtu")
    @CliName(name = "mtu", readonly = false)
    var mtu: Int = _
    @ZoomField(name = "endpoint")
    @CliName(name = "endpoint", readonly = false)
    var endpoint: InterfaceEndpoint = _
    @ZoomField(name = "port_type")
    @CliName(name = "port-type", readonly = false)
    var portType: DpPortType = _

    override def equals(obj: Any): Boolean = obj match {
        case interface: Interface => interface.name == name
        case _ => false
    }

    override def hashCode: Int = name.hashCode
}

File: brain/midonet-brain/src/main/java/org/midonet/brain/services/rest_api/models/Interface.java
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.midonet.brain.services.rest_api.models;

import java.net.InetAddress;
import java.util.UUID;

import javax.xml.bind.annotation.XmlRootElement;

import org.midonet.cluster.data.ZoomClass;
import org.midonet.cluster.data.ZoomEnum;
import org.midonet.cluster.data.ZoomEnumValue;
import org.midonet.cluster.data.ZoomField;
import org.midonet.cluster.models.Topology;
import org.midonet.cluster.util.IPAddressUtil;

@ZoomClass(clazz = Topology.Host.Interface.class)
@XmlRootElement
public class Interface {

    public UUID hostId;
    @ZoomField(name = "name")
    public String name;
    @ZoomField(name = "mac")
    public String mac;
    @ZoomField(name = "mtu")
    public int mtu;
    public int status;
    @ZoomField(name = "type")
    public Type type;
    public String endpoint;
    public String portType;
    @ZoomField(name = "addresses", converter = IPAddressUtil.Converter.class)
    public InetAddress[] addresses;

    @ZoomEnum(clazz = Topology.Host.Interface.Type.class)
    public enum Type {
        @ZoomEnumValue(value = "PHYSICAL")
        Physical,
        @ZoomEnumValue(value = "VIRTUAL")
        Virtual,
        @ZoomEnumValue(value = "TUNNEL")
        Tunnel,
        @ZoomEnumValue(value = "UNKNOWN")
        Unknown
    }

}
